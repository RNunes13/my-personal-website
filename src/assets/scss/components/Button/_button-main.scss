
.rn-button {
  @include reset-button;
  @include rem(padding, 10px 20px !important);
  @include rem(font-size, 16px !important);
  @include rem(border-radius, 30px !important);
  @include font-weight(semi-bold !important);
  text-transform: none !important;
  transition-property: color, border, background;
  transition-duration: $transition-duration;
  transition-timing-function: $accelerate-ease;

  &__primary {
    @include font-weight(semi-bold);
    border: 2px solid $color-accent;
    background-color: $color-accent;
    color: #fff;

    @include hover-supported() {
      background-color: #fff;
      color: $color-accent;
    }
  }
}
