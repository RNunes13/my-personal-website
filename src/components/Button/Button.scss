
@import '../../assets/scss/settings/_config-base';

.rn-button {
  @include reset-button;
  @include rem(padding, 10px 20px !important);
  @include rem(font-size, 16px !important);
  @include rem(border-radius, 30px !important);
  @include font-weight(semi-bold !important);
  display: flex;
  align-items: center;
  text-transform: none !important;
  outline: none !important;
  transition-property: color, border, background;
  transition-duration: $transition-duration;
  transition-timing-function: $accelerate-ease;

  &:disabled {
    cursor: not-allowed;
    border-color: $color-grey;
    color: $color-night-rider;
    background-color: $color-grey;

    svg {
      fill: $color-night-rider;
    }
  }

  svg {
    @include rem(margin-right, 8px);
    @include rem(width, 20px);
    @include rem(height, 20px);
    transition-property: fill;
    transition-duration: $transition-duration;
    transition-timing-function: $accelerate-ease;
  }

  &__primary {
    @include font-weight(semi-bold);
    border: 2px solid $color-accent;
    background-color: $color-accent;
    color: #fff;

    &:hover:not(:disabled) {
      background-color: #fff;
      color: $color-accent;

      svg { fill: $color-accent }
    }

    svg { fill: #fff }
  }
}
