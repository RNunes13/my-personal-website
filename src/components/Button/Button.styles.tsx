import styled from '@emotion/styled'
import { ctaBold } from 'styles/mixins/Typography'

export const Button = styled.button<{ fontColor?: string }>`
  display: flex;
  align-items: center;
  padding: ${({
    theme: {
      spaces: { space1, space2 },
    },
  }) => `${space1} ${space2}`};

  ${ctaBold()};

  border: 2px solid transparent;
  border-radius: 30px;
  text-transform: none !important;
  outline: none !important;

  cursor: pointer;
  transition-property: color, border, background;
  transition-duration: $transition-duration;
  transition-timing-function: $accelerate-ease;

  ${({ theme, color = 'primary', fontColor = 'base_0', disabled }) => `
    background-color: ${disabled ? theme.colors.base_20 : theme.colors[color]};
    color: ${disabled ? theme.colors.base_40 : theme.colors[fontColor]};
    fill: ${disabled ? theme.colors.base_40 : theme.colors[fontColor]};

    &:hover {
      filter: brightness(85%);
    }
  `}

  ${({ theme, disabled }) => disabled && `
    cursor: not-allowed;
    color: ${theme.colors.base_80};
    border-color: ${theme.colors.base_60};
    background-color: ${theme.colors.base_60};

    svg {
      fill: ${theme.colors.base_80};
    }
  `}

  svg {
    margin-right: 8px;
    width: ${({ theme }) => theme.icon.small};
    height: ${({ theme }) => theme.icon.small};
    transition-property: fill;
    transition-duration: 250ms;
    transition-timing-function: ${({ theme }) => theme.transitions.accelerate};
  }
`
