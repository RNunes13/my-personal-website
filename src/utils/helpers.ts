/**
 * Get a RGB color from the hexadecimal.
 * @param {string} hex Hex value in the format `#181923`.
 * @public
 */
export const getRgbFromHex = (hex = '') => {
  const hexCode = hex.replace('#', '')

  if (hexCode.length !== 6) {
    throw Error('Only six-digit hex colors are allowed.')
  }

  const rgbHex = hexCode.match(/.{1,2}/g) as RegExpMatchArray

  return [
    parseInt(rgbHex[0], 16),
    parseInt(rgbHex[1], 16),
    parseInt(rgbHex[2], 16),
  ]
}
