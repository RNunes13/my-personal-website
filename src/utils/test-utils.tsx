import { FC, ReactElement } from 'react'
import { ThemeProvider } from '@emotion/react'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import renderer, { ReactTestRenderer } from 'react-test-renderer'
import userEvent from '@testing-library/user-event'

import { theme } from 'styles/theme/Theme'

interface IProviders {
  children: ReactElement
}

const Providers: FC<IProviders> = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const renderWithTheme = (component: ReactElement): ReactTestRenderer =>
  renderer.create(<Providers>{component}</Providers>)

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(ui, { wrapper: Providers, ...options })

const renderWithUser = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => ({
  user: userEvent.setup(),
  ...render(ui, { wrapper: Providers, ...options }),
})

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { customRender as render }
export { renderWithUser }
