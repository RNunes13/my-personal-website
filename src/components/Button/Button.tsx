import { FC, ReactNode } from 'react'
import { IColorsProps } from 'styles/theme/Theme'
import { Button as ButtonStyled } from './Button.styles'

type ColorType = keyof IColorsProps

export interface ButtonProps {
  color?: ColorType
  fontColor?: ColorType
  disabled?: boolean
  className?: string
  ariaLabel?: string
  children?: string | ReactNode
  onClick?(): void
}

export const Button: FC<ButtonProps> = ({
  disabled,
  className,
  color = 'primary',
  fontColor = 'base_0',
  ariaLabel = 'Button',
  children,
  onClick,
}) => {
  if (!children) return null

  return (
    <ButtonStyled
      type="button"
      color={color as string}
      fontColor={fontColor as string}
      disabled={disabled}
      className={className}
      aria-label={ariaLabel}
      onClick={onClick}
      data-testid='button'
    >
      {children}
    </ButtonStyled>
  )
}

export default Button
