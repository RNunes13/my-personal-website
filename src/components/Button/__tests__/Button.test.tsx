import { render, renderWithUser, renderWithTheme } from 'utils/test-utils'
import { Button } from '../Button'
import type { ButtonProps } from '../Button'

const defaultProps: ButtonProps = {
  children: 'Button label',
}

const getComponent = (props: Partial<ButtonProps> = {}) => (
  <Button {...defaultProps} {...props} />
)

describe('Button unit test', () => {
  it('should call the provide click functon', async () => {
    const onClick = jest.fn()
    const { getByTestId, user } = renderWithUser(getComponent({ onClick }))
    const button = getByTestId('button')

    expect(onClick).not.toHaveBeenCalled()

    await user.click(button)

    expect(onClick).toHaveBeenCalled()
  })

  it('should render nothing if no children is provided', () => {
    const { container } = render(getComponent({ children: null }))
    expect(container).toBeEmptyDOMElement()
  })

  it('should render the button with the primary color', () => {
    const tree = renderWithTheme(getComponent({ color: 'primary' })).toJSON()
    expect(tree).toMatchSnapshot('Button - Primary color')
  })

  it('should render the button with the accent color', () => {
    const tree = renderWithTheme(getComponent({ color: 'accent' })).toJSON()
    expect(tree).toMatchSnapshot('Button - Accent color')
  })
})
