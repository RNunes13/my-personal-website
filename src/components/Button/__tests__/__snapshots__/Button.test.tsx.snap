// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button unit test should render the button with the accent color: Button - Accent color 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 20px;
  font-family: Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 700;
  font-size: 14px;
  line-height: 14px;
  border: 2px solid transparent;
  border-radius: 30px;
  text-transform: none!important;
  outline: none!important;
  cursor: pointer;
  transition-property: color,border,background;
  transition-duration: $transition-duration;
  transition-timing-function: $accelerate-ease;
  background-color: #419866;
  color: #FFFFFF;
  fill: #FFFFFF;
}

.emotion-0:hover {
  -webkit-filter: brightness(85%);
  filter: brightness(85%);
}

.emotion-0 svg {
  margin-right: 8px;
  width: 16px;
  height: 16px;
  transition-property: fill;
  transition-duration: 250ms;
  transition-timing-function: cubic-bezier(0.4, 0.0, 1, 1);
}

<button
  aria-label="Button"
  className="emotion-0 emotion-1"
  color="accent"
  data-testid="button"
  type="button"
>
  Button label
</button>
`;

exports[`Button unit test should render the button with the primary color: Button - Primary color 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px 20px;
  font-family: Roboto,Oxygen,Ubuntu,Cantarell,"Fira Sans","Droid Sans",BlinkMacSystemFont,-apple-system,sans-serif;
  font-style: normal;
  font-weight: 700;
  font-size: 14px;
  line-height: 14px;
  border: 2px solid transparent;
  border-radius: 30px;
  text-transform: none!important;
  outline: none!important;
  cursor: pointer;
  transition-property: color,border,background;
  transition-duration: $transition-duration;
  transition-timing-function: $accelerate-ease;
  background-color: #181923;
  color: #FFFFFF;
  fill: #FFFFFF;
}

.emotion-0:hover {
  -webkit-filter: brightness(85%);
  filter: brightness(85%);
}

.emotion-0 svg {
  margin-right: 8px;
  width: 16px;
  height: 16px;
  transition-property: fill;
  transition-duration: 250ms;
  transition-timing-function: cubic-bezier(0.4, 0.0, 1, 1);
}

<button
  aria-label="Button"
  className="emotion-0 emotion-1"
  color="primary"
  data-testid="button"
  type="button"
>
  Button label
</button>
`;
